readme.md

```diff
@@ -1,6 +1,264 @@
+* Simulate a review process by reviewing github pull requests
+  and addressing issues identified there
+* Include better project and code documentation for ease of use
+
+## Project progress
+
+My original modifications included 5 major changes:
+
+* pull-from-git
+  * Allow updates to the autograder to be dynamically pulled via git
+* required-files
+  * Automate filename copying and checking (with feedback to students)
+* time-limited
+  * Integrate time limits by running tests in their own processes, redirecting stdout/stderr
+* fcn-tests
+  * Automate scanning a matrix of parameters in tests, particularly for numerical functions
+* correct-with-git
+  * Create a framework for grader to incrementally correct students' code using git,
+    preventing minor mistakes from disproportionately affecting grades
+
+So far, the first two changes have been merged into this project,
+and the remaining are in development
+(likely with some code being pushed on branches on this repo,
+or even with open pull requests).
+
+### pull-from-git
+
+Gradescope's default autograder configuration workflow requires
+the grader to upload the autograder in a zip file,
+which Gradescope then uses to build a Docker image --
+an image which will then be used to spawn an autograder instance
+every time the autograder is run on a students' submission.
+The problem is that this build process takes a few minutes,
+and this Docker image needs to be rebuilt every time
+the grader wishes to make a change to the autograder.
+This often creates a frustrating bottleneck in the debugging process.
+
+With this pull-from-git enchancement,
+we can circumvent the need to rebuild the autograder image
+by downloading updates dynamically via git.
+Every time the autograder is run on a student's assignment submission,
+the autograder checks for any updates that have been pushed to the
+remote repository since the autograder was configured
+(i.e., since the project directory was zipped and uploaded to Gradescope's
+website, and the setup script was run).
+The autograder uses the settings in the project's `.git` directory
+to pull any such updates.
+
+
+Notes:
+
+* This type of enhancement was explicitly
+  [suggested][gs-docs-pull-from-git]
+  in the Gradescope's autograder docs,
+  and my code is partially based on their
+  [example code][gs-gh-pull-from-git-ex].
+* The update process is performed
+  not only just before the autograder is run on a student's submission,
+  but also just after
+  the autograder is uploaded to Gradescope -- when `setup.sh` is run
+* The autograder runs `git reset --hard` before doing this,
+  which is a workaround for Windows users.  This means that any changes to the
+  autograder *must* be committed (using `git commit`) before being uploaded
+  (see [issue \#2][issue-2])
+  * Once can disable this reset by removing or commenting out
+    the `git reset --hard` line from `update.sh`
+
+
+Inner workings:
```

Comment:
This "inner workings" section could probably be more complete/ self contained.  Perhaps much of the previous list could be merged into it?

In any case, it's probably a good idea to separate a high-level explanation of how it works with nitty-gritty implementation and usage details